### PR TITLE
Added ability to set arbitrary http headers on batch requests.

### PR DESCRIPTION
**Pull request by nsarris**
Add the abbility to set http headers on batch operations. Changes made to achieve this:

- Added WithHeader/WithHeaders methods to ODataBatch
- Added GetRequest method to ODataBatch to match GetRequestFor functionality on single operations
- ODataRequest now internally carries the batch state (actions and response indeces) needed for replaying actions to set closure values. This is also usefully for exposing the GetRequest.

### CODE REVIEW

File: src/Simple.OData.Client.Core/Adapter/BatchWriterBase.cs [hunk 1]
```diff
@@ -27,9 +28,12 @@ protected BatchWriterBase(ISession session, IDictionary<object, IDictionary<stri
         public abstract Task<HttpRequestMessage> EndBatchAsync();
 
         public async Task<ODataRequest> CreateBatchRequestAsync(
-            IODataClient client, IList<Func<IODataClient, Task>> actions, IList<int> responseIndexes)
+            IODataClient client, IList<Func<IODataClient, Task>> actions,
```
Comment (watfordgnf) [inline src/Simple.OData.Client.Core/Adapter/BatchWriterBase.cs:31]: Is there a reason we need to remove `responseIndexes` from a public interface? That's a potentially big breaking change for some folks.

File: src/Simple.OData.Client.Core/ODataClient.Async.cs [hunk 2]
```diff
@@ -1120,12 +1120,23 @@ internal async Task<T[]> ExecuteAsArrayAsync<T>(FluentCommand command, Cancellat
                 : result.Select(x => (T)x.ToObject(_session.TypeCache, typeof(T))).ToArray();
         }
 
-        internal async Task ExecuteBatchAsync(IList<Func<IODataClient, Task>> actions, CancellationToken cancellationToken)
+        internal async Task<ODataRequest> BuildRequestForBatch(IList<Func<IODataClient, Task>> actions, CancellationToken cancellationToken, IDictionary<string,string> headers = null)
```
Comment (watfordgnf) [inline src/Simple.OData.Client.Core/ODataClient.Async.cs:1123]: Since this is internal, you should make `headers` non-optional and order it before the cancellation token.

File: src/Simple.OData.Client.Core/ODataClient.Async.cs [hunk 3]
```diff
@@ -1120,12 +1120,23 @@ internal async Task<T[]> ExecuteAsArrayAsync<T>(FluentCommand command, Cancellat
...
             if (cancellationToken.IsCancellationRequested) cancellationToken.ThrowIfCancellationRequested();
 
-            await ExecuteBatchActionsAsync(actions, cancellationToken).ConfigureAwait(false);
+            if (!actions.Any())
+                return null;
+
+            return await _lazyBatchWriter.Value.CreateBatchRequestAsync(this, actions, headers).ConfigureAwait(false);
+        }
+
+        internal async Task ExecuteBatchAsync(IList<Func<IODataClient, Task>> actions, CancellationToken cancellationToken, IDictionary<string, string> headers = null)
```
Comment (watfordgnf) [inline src/Simple.OData.Client.Core/ODataClient.Async.cs:1134]: Since this is internal, you should make `headers` non-optional and order it before the cancellation token.

File: src/Simple.OData.Client.Core/ODataClient.Async.cs [hunk 4]
```diff
@@ -1120,12 +1120,23 @@ internal async Task<T[]> ExecuteAsArrayAsync<T>(FluentCommand command, Cancellat
...
         }
 
-        internal async Task ExecuteBatchAsync(IList<Func<IODataClient, Task>> actions, CancellationToken cancellationToken)
+        internal async Task<ODataRequest> BuildRequestForBatch(IList<Func<IODataClient, Task>> actions, CancellationToken cancellationToken, IDictionary<string,string> headers = null)
         {
             await _session.ResolveAdapterAsync(cancellationToken).ConfigureAwait(false);
             if (cancellationToken.IsCancellationRequested) cancellationToken.ThrowIfCancellationRequested();
 
-            await ExecuteBatchActionsAsync(actions, cancellationToken).ConfigureAwait(false);
+            if (!actions.Any())
```
Comment (watfordgnf) [inline src/Simple.OData.Client.Core/ODataClient.Async.cs:1128]: Can we avoid an unrelated behavioral change like this? Not saying it isn't a good idea, but preferably part of a separate PR that could be reverted on its own if it impacts something.

File: src/Simple.OData.Client.Core/ODataRequest.cs [hunk 5]
```diff
@@ -57,7 +58,13 @@ public string[] Accept
...
         public bool ResultRequired { get; set; }
         public bool CheckOptimisticConcurrency { get; set; }
-        public readonly IDictionary<string, string> Headers = new Dictionary<string, string>();
+        public IDictionary<string, string> Headers { get; } = new Dictionary<string, string>();
+
+        /// <summary>
+        /// Internal state used for batch requests. Needed in order to replay operations to assign return values to closures.
+        /// </summary>
+        internal IList<int> BatchResponseIndeces { get; set; } = new List<int>();
+        internal IList<Func<IODataClient, Task>> BatchActions { get; set; } = new List<Func<IODataClient, Task>>();
```
Comment (watfordgnf) [inline src/Simple.OData.Client.Core/ODataRequest.cs:67]: Why was this needed as part of this work? (curious)

If its a perf gain, it would be better in another PR.

**Review by watfordgnf — COMMENTED** [review]
Few questions and comments.

File: src/Simple.OData.Client.Core/ODataRequest.cs (hunk 5 above)
Comment (nsarris) [reply src/Simple.OData.Client.Core/ODataRequest.cs:67]: This is a fix already commited and merge, I will remove it from this PR.

Comment (nsarris) [reply src/Simple.OData.Client.Core/ODataRequest.cs:67]: The intent behind the seemlingly unrelated functional changes was the ability to test the implementation. For all the singular operations you can get the request using the GetRequestFor() API. There is no such API for batch and digging deeper I found out that it's impossible to return an ODataRequest and then send the request without breaking the feature with the returning closures. I think it's much more sound architecturally, to encapsulate this "state" of actions and indeces inside the request object and pass it on for execution. So I can revert these changes and just commit the header setting part but there will simply be no way to write an end to end test for this feature.

**Comment by nsarris** [comment]
@watfordgnf Please see my comment above about all the functional changes. I can provide a PR without them and then provide a PR on top if you like to take if from there.

File: src/Simple.OData.Client.Core/ODataRequest.cs (hunk 5 above)
Comment (watfordgnf) [reply src/Simple.OData.Client.Core/ODataRequest.cs:67]: I'm not familiar enough with this pathway to know how big of an impact this change is, perhaps @object  or @phatcher  can weigh in.

It seems reasonable on face value (with the minor nit to change `Indeces` to `Indices` or `Indexes`).

Comment (nsarris) [reply src/Simple.OData.Client.Core/ODataRequest.cs:67]: Indeces is a type on my part, it was supposed to be Indices. It's minor I can revert to indexes it's internal so it doesn't break anything.

**Comment by nsarris** [comment]
@watfordgnf I found away to write an end to end test, I will close this one and create a new PR without the functional changes.
